### PR TITLE
Cache results dont affect average execution time

### LINKED ISSUE #15432: Cached query responses incorrectly update query.average_execution_time
It cause by wrong calculation of "query.average_execution_time".
I found that average_execution_time is including cache call time.

For example when I run long query, at the first time, it goes to cache.
And next run, the result comes from the cache and recalculate average_execution_time.
As time series like this(cache threshold is 10 seconds):
1. First run
   - run time: 1min
   - average_execution_time: 1min (this will be cached)
2. Second run
   - run time: 1s (cause it came from cache)
   - average_execution_time: 30.5s
3. Third run
   - run time: 1s (cause it came from cache)
   - average_execution_time: 15.75s
4. Fourth run
   - run time: 1s (cause it came from cache)
   - average_execution_time: 8.375s
5. Fifth run
   - run time: 1min (cause average_execution_time is less than threshold)
   - average_execution_time: 34s (this will be cached again)

This problem makes cache policy useless.
If two or more person sees same dashboard, query executions are goes to nightmare.
"average_execution_time" needs to calculate only actual runs.

### PR DESCRIPTION
**Pull request by dpsutton**
Fixes #15432 

Fix is quite simple, don't update the execution times for cached results. Noticed that we are calling the completion arity twice and we were actually adding two cached execution times into the average, making it decrease even faster than just single counting the cached time.

My eyes glazed over the call of the single completion arity from the step arity for a long time. I thought it was related to the cache-miss codepath doing work after a cache-hit process as well.

### CODE REVIEW

File: src/metabase/models/query.clj [hunk 1]
```diff
@@ -13,7 +13,8 @@
 (u/strict-extend (class Query)
   models/IModel
   (merge models/IModelDefaults
-         {:types (constantly {:query :json})}))
+         {:types       (constantly {:query :json})
+          :primary-key (constantly :query_hash)}))
```
Comment (dpsutton) [inline src/metabase/models/query.clj:17]: needed so we can delete the `Query` record (if any) in the test

File: src/metabase/query_processor/middleware/cache.clj [hunk 2]
```diff
@@ -139,8 +139,7 @@
 
         ([acc row]
          (if (map? row)
-           (do (vreset! final-metadata row)
-               (rf acc))
```
Comment (dpsutton) [inline src/metabase/query_processor/middleware/cache.clj:143]: took quite a while to realize we were doing this here. We can just not process this row and save it in the final-metadata

File: test/metabase/query_processor/middleware/cache_test.clj [hunk 3]
```diff
@@ -323,7 +327,32 @@
                        (-> cached-result
                            (dissoc :cached :updated_at)
                            (m/dissoc-in [:data :results_metadata :checksum])))
-                    "Cached result should be in the same format as the uncached result, except for added keys")))))))))
+                    "Cached result should be in the same format as the uncached result, except for added keys"))))))))
+  (testing "Cached results don't impact average execution time"
```
Comment (howonlee) [inline test/metabase/query_processor/middleware/cache_test.clj:331]: nit: space

Comment (dpsutton) [reply test/metabase/query_processor/middleware/cache_test.clj:331]: space where? I'm not following.

File: test/metabase/query_processor/middleware/cache_test.clj [hunk 4]
```diff
@@ -323,7 +327,32 @@
...
+      (with-redefs [process-userland-query/save-query-execution!* (fn [& args]
+                                                                    (swap! call-count inc)
+                                                                    (apply save-query-execution-original args)
+                                                                    (deliver called-promise true))
+                    cache/min-duration-ms                         (constantly 0)]
+        (with-mock-cache [save-chan]
+          (db/delete! Query :query_hash q-hash)
+          (is (not (:cached (qp/process-userland-query query (context.default/default-context)))))
+          (a/alts!! [save-chan (a/timeout 200)]) ;; wait-for-result closes the channel
+          (u/deref-with-timeout called-promise 500)
```
Comment (howonlee) [inline test/metabase/query_processor/middleware/cache_test.clj:347]: this one flaked for me over three tries testing.

**Review by howonlee — APPROVED** [review]
tests run fine, actual functionality seems to be fine, I encountered a flake once, dunno if you wanna whack it tho

File: test/metabase/query_processor/middleware/cache_test.clj (hunk 4 above)
Comment (dpsutton) [reply test/metabase/query_processor/middleware/cache_test.clj:347]: can you try finding better values for the `200` and `500` numbers? I can see if there is some decent way to make it more reliable. But fundamentally I need things to exist in a cache, and have timings in the db, both of which happen asynchronous to the query thread.

File: test/metabase/query_processor/middleware/cache_test.clj (hunk 3 above)
Comment (howonlee) [reply test/metabase/query_processor/middleware/cache_test.clj:331]: actually, damn. there's a space between each `testing` clause in all the api tests and this is very inconsistent within QP tests. I guess we should just follow the format within each file or whack it entirely per file, not piecemeal like this, so never mind on this one

Comment (dpsutton) [reply test/metabase/query_processor/middleware/cache_test.clj:331]: can you link to an example? I'm not following where a space would logically go?

Comment (howonlee) [reply test/metabase/query_processor/middleware/cache_test.clj:331]: https://github.com/metabase/metabase/blob/9587fdbc4e98b06511575fc5dfb60328fbb7957e/test/metabase/api/collection_test.clj#L73

File: test/metabase/query_processor/middleware/cache_test.clj (hunk 4 above)
Comment (howonlee) [reply test/metabase/query_processor/middleware/cache_test.clj:347]: I tried 10 times at `200` and `800` in the same settings and it hasn't flaked yet

File: test/metabase/query_processor/middleware/cache_test.clj (hunk 3 above)
Comment (dpsutton) [reply test/metabase/query_processor/middleware/cache_test.clj:331]: ah newlines. i see. asking around if we have a standard. I don't care one way or the other.

Comment (howonlee) [reply test/metabase/query_processor/middleware/cache_test.clj:331]: memorializing side channel dealio: "season to taste"

**Pull request merged**